wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -1652,4 +1652,76 @@ private class WasmExpressionBuilder private (
     instrs += ARRAY_NEW_FIXED(TypeIdx(arrTy.name), I32(t.elems.size))
     t.tpe
   }
+
+  private def genClosure(tree: IRTrees.Closure): IRTypes.Type = {
+    implicit val ctx = this.ctx
+
+    val hasThis = !tree.arrow
+    val dataStructType = ctx.getClosureDataStructType(tree.captureParams.map(_.ptpe))
+
+    // Define the function where captures are reified as a `__captureData` argument.
+    val closureFuncName = fctx.genInnerFuncName()
+    locally {
+      val receiverParam =
+        if (!hasThis) None
+        else Some(WasmLocal(WasmLocalName.receiver, Types.WasmAnyRef, isParameter = true))
+
+      val captureDataParam = WasmLocal(
+        WasmLocalName("__captureData"),
+        Types.WasmRefType(Types.WasmHeapType.Type(dataStructType.name)),
+        isParameter = true
+      )
+
+      val paramLocals = (tree.params ::: tree.restParam.toList).map { param =>
+        val typ = TypeTransformer.transformType(param.ptpe)
+        WasmLocal(WasmLocalName.fromIR(param.name.name), typ, isParameter = true)
+      }
+      val resultTyps = TypeTransformer.transformResultType(IRTypes.AnyType)
```

Comment:
Because the declared result type is implicitly always `any`. If the body happens to be a primitive `int`, we need to box it to match the declared result type. I expanded a little bit on the comment on `generateIRBody`.